### ISSUE TITLE
Random behavior and incorrect gradient

### ISSUE THREAD

**Issue by mojtabamozaffar** [taichi-dev/taichi#1594]
Hi,

Consider a case where you have a differentiable simulation response and you want to optimize the amount of time that the respond spends in a specific range. Here is a simple example of it:

```
np.random.seed(1)

real = ti.f32
intg = ti.i32
ti.init(default_fp=real)

loss = ti.var(dt=real)

coeficient_1 = ti.var(dt=real, needs_grad=True)
coeficient_2 = ti.var(dt=real)

result = ti.var(dt=real)
time_spend_in_range = ti.Vector(2, dt=real)

is_in_range = ti.var(dt=real)

endtime = 10.0
delta_t = 0.1
num_steps = int(endtime/dt)
coeficient_1_init = 2.0
coeficient_2_init = 0.4
minimum_range = 0.05
maximum_range = 0.15

ti.root.dense(ti.l, num_steps).place(result)
ti.root.place(loss, coeficient_1, coeficient_2, time_spend_in_range, is_in_range)
ti.root.lazy_grad()

@ti.kernel
def clear_grads():
    coeficient_1.grad[None] = 0.0  
    
def initialize():
    coeficient_1[None] = coeficient_1_init
    coeficient_2[None] = coeficient_2_init
    
# calculating result for every time step according to a formula output = exp(-at)*sin(t)
@ti.kernel
def time_integrate(t: intg, dt: real):
        result[t] = ti.exp(-1* coeficient_2[None] * t * dt) * ti.sin(coeficient_1[None] * t * dt)

# detect the last time result passes from maximum_range to minimum_range and record the corresponsing 
# times in time_spend_in_range[None][0] and time_spend_in_range[None][1] respectively
@ti.kernel
def detect_range():
    for t in range(1, num_steps):
        if result[t] <= maximum_range and result[t-1] >= maximum_range:
            time_spend_in_range[None][0] = t*dt - (result[t] - maximum_range)*dt/(result[t]-result[t-1])
            is_in_range[None] = 1.0
        if result[t] <= minimum_range and result[t-1] >= minimum_range and is_in_range[None] > 0.5:
            time_spend_in_range[None][1] = t*dt - (result[t] - minimum_range)*dt/(result[t]-result[t-1])
            is_in_range[None] = 0.0        

# defining the loss to maximum the time spent in the range
@ti.kernel
def compute_loss():
    ti.atomic_add(loss, -1 * (time_spend_in_range[None][1]-time_spend_in_range[None][0]))
            
def simulate():
    for time_step in range(0, num_steps):
        time_integrate(time_step, dt)
    detect_range()
    compute_loss()
```

If I run it using the following script and plot it everything looks good. It calculates the time spent between two ranges (you can see the times as dashed vertical lines in the figure below) as well as the gradient with respect to the coefficient.

```
loss[None] = 0.0
initialize()
clear_grads()
with ti.Tape(loss):
    simulate()
result_np = result.to_numpy()
coeficient_1.grad[None]
```
![fig_1](https://user-images.githubusercontent.com/55418576/88470280-20f21480-cec0-11ea-964d-3f293cea984e.png)

But the first problem appears if I try to run the code multiple times. If I try to calculate the results for 10 different coefficient values and repeat the exact process 4 times, I get a different result every time. Here is the script I used:

```
n_tests = 10
fig, axs = plt.subplots(4, figsize=(6,8))
for repeat in range(4):
    coeficient_1_inits = np.linspace(1.95, 2.05, n_tests)
    losses = []
    for i in range(n_tests):
        coeficient_1_init = coeficient_1_inits[i]
        loss[None] = 0.0
        initialize()
        clear_grads()
        with ti.Tape(loss):
            simulate()
        losses.append(loss[None])
    axs[repeat].plot(coeficient_1_inits,losses)
plt.tight_layout()
plt.show()
```

![fig_2](https://user-images.githubusercontent.com/55418576/88470316-91009a80-cec0-11ea-9913-c89f2361caea.png)

Particularly it seems one or two samples randomly result in incorrect values. The reason why I think it is incorrect is that (1) the values of coeficient_1 are very close (in the range of 1.95 and 2.05) and I don't expect any sharp change in that range and (2) it appears randomly at a different place each time.

The other problem is that (neglecting sharp changes) it can be seen that the response and coefficient_1 have a positive correlation and therefore the gradient of response with respect to coefficient_1 at point 2.0 should be logically positive. However, the reported gradient is negative (-2.28), which doesn't make sense. If I optimize coefficient_1 according to the computed gradients the loss will continuously increase.

I have been going over the code multiple times to make sure it is not a bug in my code and couldn't find any bug in it so far. Any input or suggestion is appreciated!

**Comment by ehannigan** [comment]
Just to back up @mojtabamozaffar, I had the same reproducibility when running mass_spring.py (one of the difftaichi examples). Even when I remove the random initialization of the neural network weights, the resulting loss changes for each run. I posted this in issue #1565. If you want me to run this and post results, let me know.
